Replicator/ReplicatorOptions.hh

```diff
@@ -315,26 +329,89 @@ namespace litecore { namespace repl {
             back.pullFilter = collOpts.pullFilter;
             back.callbackContext = collOpts.callbackContext;
         }
-        trimDisabledCollections();
     }
 
-    inline void Options::trimDisabledCollections() {
+    inline void Options::verify() const {
```

Comment:
One thing that feels little weird is that without calling verify() some status (e.g. _isActive and _collectionAware) is not valid.

Reply:
@pasin this is because Options can be used as a data structure outside Replicator. I saw in code that we create an Options with both push/pull disabled which caused my initial attempt to fail. I later decided to enforce the sanity of if it is used in Replicator. Hence, verify() becomes a public method as opposed to doing it in the Options' constructor. 
I prefer not to use optional. Using optional sends a message to the called that every time using it, he/she needs to test it. I want to hint a semantics that isActive, collectionAware, etc. are attributes that are required for Replicators, and therefore, before reading them, code must verify them. Currently, failure of verify() throws error right away.

Reply:
Well it seems like you are just lying if you put it that way.  You say you want to avoid "needing to test every time" but the "fix" here is to just pretend that it is always ok when it is not.  There is an actual tri-state value here unless you want to rework the way this happens.  I can see a method that returns a COPY of the options from verify that is actually completely immutable instead of just sort of frozen halfway through.

Reply:
I think this representation is clear: when used by Replicator, and all related workers, some constraints apply. As the head of group of workers, Replicator should call verify() to ensure the constraints. As by-products, with the constraints being satisfied, some derived properties can be defined, such as isActive. With the constraints being satisfied, isActive is not well defined, because some collection is active, some can be disabled, and yet some can be passive. 
Returning a copy indeed can maintain the mutability, but there is a practicality concern. The current architecture ensures that all workers are created with the shared Options object. For example, once inside Replicator's constructor, Connection, BLIPIO, etc., have already received shared Options. It's awkward to swap the Options for all of them. If isActive is an c++ optional, then, we either have to structurally guarantee that it contains a value after certain point, or every time we have to say
if (!isActive.has_value()) throw runtime logical error. else if (*isActive) { } else { }